### DISCUSSION TITLE
Can I use varibles to define the topic in flow message source?

### DISCUSSION THREAD

**Discussion by jpiwek** [emqx/emqx#13682] [General]
Can I add variables like "${client_id} in message source like this: /${client_id}/shadow/#

**Comment by zmstone** [comment]
Hi @jpiwek 

tl;dr;
No. The MQTT data source topic does not support template rendering.
This is because the subscriptions are not as dynamic as message publish, subscriptions are sent when the ingress bridge starts up.
Also, there is no variable context during create/start-up.

## Understand how MQTT data source works:

Also, judging by the example you provided `/${client_id}/shadow/#`, (btw, is there really a leading `/`?) you maybe have a misunderstood how EMQX MQTT data source works: EMQX starts a pool of MQTT clients to subscribe a remote broker, the clients has their client ID prefix configurable. i.e. `${client_id}` would be the configured prefix + a random suffix.

If you want to replicate `/{client-id}/shadow/#` messages from a remote broker/cluster, you should configure the source with
`$share/{share-sub-group-name}//+/shadow/#` (NOTE: `//` intended, assuming you have a leading `/` in the remote topics)
And then add a `republish` action to publish the messages to local MQTT clients.

## More to consider

If the remote MQTT broker supports data bridge like EMQX (or maybe it is EMQX?), an alternative is to make the remote brokers publish instead of local (EMQX) to subscribe. --- and I guess the remote should publish with retain flag set if it's a shadow object.

If both ends are EMQX, you may consider EMQX Enterprise cluster-liking feature (will be released in 5.8), which is optimized to forward messages only when there are MQTT clients subscribed to the topics, but not to send ALL messages matching the filter.

**Comment by jpiwek** [comment]
Ok thanks for your feedback. What I actually have in mind, is to create an example which takes EMQX MQTT data and forwards it to e.g. Supabase for realtime event handling and data storage.
I tried a basic example with Flow Designer, which worked. On another thread Webhooks were proposed as well. Will also give it a try.

As stated above. At the moment what I would like to achieve is to create kind of shadow for MQTT topics with flow/webhook forwarder  to e.g. Supabase.
One shadow is supposed to store IoT device telemetry, device configraution  data 
and another one to automcatically register devices in Supabase database using public certificate/public key or JWT (like GCP Core did in the past).

The approach I have in mind is similar to AWS IoT Core service.
The question is what is the best way to structure shadow topics (for multiple IoT device user/owners) and MQTT configuration (security to access shadow per user/owner) with EMQX in kind of n:m relation?

Thanks,
